text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,439 @@
+        │   │   └── t
+        │   │       └── t-desc-0-.js
+        │   ├── settings.html
+        │   ├── src
+        │   │   └── t
+        │   │       └── lib.rs.html
+        │   └── t
+        │       ├── all.html
+        │       ├── index.html
+        │       ├── sidebar-items.js
+        │       └── trait.T.html
+        ├── doc.parts
+        │   └── t
+        │       └── crate-info.json
+        └── libt.rmeta
+</pre>
+
+</details>
+
+## Suggested workflows
+
+### No cross-crate information 
+Provide `--merge=none` to every invocation of rustdoc.
+
+### Cross-crate information, mutate shared directory
+
+Use `--merge=read-write`, and specify the same `--out-dir` to every invocation of rustdoc. `--merge=read-write` will be the default value if `--merge` is not provided. This is the workflow that Cargo uses.
+
+### Cross-crate information, no shared directory
+
+Specify a different `--out-dir` to every invocation of rustdoc. Additionally, you should provide `--write-info-json=<crate-specific path to file>` and `--merge=none` when documenting the dependencies of your root crate. Then, when you document the root crate, you will provide  `--include-info-json=<crate-specific path selected previously>`, `--include-rendered-docs=<out dirs chosen previously>` for each one of your dependencies, and `--merge=write-only`. You should provide `--extern-html-root-url`, and specify a static, absolute location for the URL. This flag, with the same URL, will be needed for every invocation of rustdoc, for every dependency.
+
+# Reference-level explanation
+
+The existing cross-crate information files, like `search-index.js`, all are lists of elements, rendered in an specified way (e.g. as a JavaScript file with a JSON array or an HTML index page containing an unordered list). The current rustdoc (in `write_shared`) pushes the current crate's version of the CCI into the one that is already found in `doc`, and renders a new version. The rest of the proposal uses the term **part** to refer to the pre-merged, pre-rendered element of the CCI. This proposal does not add any new CCI or change their contents (modulo sorting order, whitespace).
+
+## New file: `crate-info.json`
+
+`crate-info.json` is an artifact that encodes the partial contents and destination of several cross-crate information files. It only encodes information about a single-crate. This file is written if `--write-info-json` is provided. The current crate's information and any `crate-info.json` added through `--include-info-json` are merged and rendered if `--merge=read-write` or `--merge=write-only` are provided.
+
+The content of `crate-info.json` is unstable. Rustdoc only guarantees that it will accept `crate-info.json` files written by the same version of rustdoc, and rustdoc is the only explicitly supported consumer of `crate-info.json`. Only the presence of `crate-info.json` is stabilized. Non-normatively, there are several pieces of information that `crate-info.json` may contain:
```

Comment:
Do we strictly *need* to stabilize even the presence of `crate-info.json`? It seems like we should only stabilize the existence of *some* "parts" directory whose contents are opaque and the fact that rustdoc will accept a "parts" directory produce by the same version of rustdoc. In other words, bring it up a level and only specify/stabilize the directory rather than the (currently sole, but in the future maybe one of several) file within it.

Reply:
I agree with you, and I will make this change.

Reply:
That's an important concern! I will mention that, and the fact that the directory structure may change when the edition changes.

Reply:
> We kind of do, most custom build systems need to explicitly declare which files are intended to "escape containment".

Hmm, is it not possible to specify that an entire directory escapes containment? Or use a glob within that directory?

If those are not possible, then I would still prefer not to require an edition change for the directory's structure to change.

Reply:
I removed the edition requirement and said that rustdoc will make an effort to use a crate-info file, but the contents are formally unstable.